src/Features/LanguageServer/Protocol/Handler/Diagnostics/AbstractDocumentPullDiagnosticHandler.cs

```diff
@@ -2,23 +2,61 @@
 
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
 using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.LanguageServer.Handler.Diagnostics.DiagnosticSources;
 using Microsoft.CodeAnalysis.Options;
 using Microsoft.CommonLanguageServerProtocol.Framework;
 using Roslyn.LanguageServer.Protocol;
-using LSP = Roslyn.LanguageServer.Protocol;
 
 namespace Microsoft.CodeAnalysis.LanguageServer.Handler.Diagnostics;
 
 internal abstract class AbstractDocumentPullDiagnosticHandler<TDiagnosticsParams, TReport, TReturn>(
     IDiagnosticAnalyzerService diagnosticAnalyzerService,
     IDiagnosticsRefresher diagnosticRefresher,
+    IDiagnosticSourceManager diagnosticSourceManager,
     IGlobalOptionService globalOptions)
     : AbstractPullDiagnosticHandler<TDiagnosticsParams, TReport, TReturn>(
         diagnosticAnalyzerService,
         diagnosticRefresher,
         globalOptions), ITextDocumentIdentifierHandler<TDiagnosticsParams, TextDocumentIdentifier?>
     where TDiagnosticsParams : IPartialResultParams<TReport>
 {
-    public abstract LSP.TextDocumentIdentifier? GetTextDocumentIdentifier(TDiagnosticsParams diagnosticsParams);
+    protected readonly IDiagnosticSourceManager DiagnosticSourceManager = diagnosticSourceManager;
+
+    public abstract TextDocumentIdentifier? GetTextDocumentIdentifier(TDiagnosticsParams diagnosticsParams);
+
+    protected override ValueTask<ImmutableArray<IDiagnosticSource>> GetOrderedDiagnosticSourcesAsync(TDiagnosticsParams diagnosticsParams, string? requestDiagnosticCategory, RequestContext context, CancellationToken cancellationToken)
+    {
+        if (GetOpenDocument(context) is null)
+            return new([]);
+
+        return DiagnosticSourceManager.CreateDiagnosticSourcesAsync(context, requestDiagnosticCategory, isDocument: true, cancellationToken);
+    }
+
+    protected static TextDocument? GetOpenDocument(RequestContext context)
+    {
+        // Note: context.Document may be null in the case where the client is asking about a document that we have
+        // since removed from the workspace.  In this case, we don't really have anything to process.
+        // GetPreviousResults will be used to properly realize this and notify the client that the doc is gone.
+        //
+        // Only consider open documents here (and only closed ones in the WorkspacePullDiagnosticHandler).  Each
+        // handler treats those as separate worlds that they are responsible for.
+        var textDocument = context.TextDocument;
+        if (textDocument is null)
+        {
+            context.TraceInformation("Ignoring diagnostics request because no text document was provided");
+            return null;
+        }
+
+        if (!context.IsTracking(textDocument.GetURI()))
+        {
+            context.TraceWarning($"Ignoring diagnostics request for untracked document: {textDocument.GetURI()}");
+            return null;
+        }
```

Comment:
did this code move here from somewhere else?

Reply:
why is this not using context.GetTrackedDocument?

Reply:
yes